sycl/doc/extensions/Barrier/Barrier.asciidoc

```diff
@@ -0,0 +1,184 @@
+= SYCL_EXT_ONEAPI_BARRIER
```

Comment:
For now this extension is implemented only for CUDA. However, I don't think it is in general impossible to implement this for other backends. I have given it some thought and at least I have an idea how this could be implemented for lower sm versions that do not have hardware support for this. So I am not sure if we want to name this as a CUDA specific extension.

Reply:
Maybe I was unclear in the description of this PR. Now I clarified that this is not intended to be CUDA-only extension, just that for now we only have CUDA implementation.

Reply:
I remain skeptical.  By tying this API to the number of work-items expected to arrive, you're permitting synchronization at a finer granularity than may be guaranteed to be supported by other hardware/backends.  For example, I can't think of a way to implement this extension on top of a generic OpenCL device, relying only on the execution model described in the OpenCL specification.

Reply:
That is true. When I was thinking about software implementation of barrier I figured I would need either independent forward progress or named barriers that can specify the number of participating work-items. Both of these features are available in CUDA, but not OpenCL. I also checked at HIP, and it looks like it supports nothing similar. 

So I think we need to determine if a common extension for CUDA, esimd and possibly level 0 (I did not look into that) makes sense.

Reply:
Exactly.  And even a named barrier might not be sufficient.  SPIR-V exposes named barriers ([OpNamedBarrierInitialize](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#OpNamedBarrierInitialize) and [OpMemoryNamedBarrier](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#OpMemoryNamedBarrier)), but these instructions are specified in terms of how many sub-groups they synchronize rather than how many work-items they synchronize.

I do think a common extension for split barriers makes sense.  But I wonder if it would make sense to start with something simpler (e.g. a barrier with exactly the same control-flow requirements and semantics as SYCL work-group barrier, but with separate `arrive` and `wait` phases).  I've not thought about this enough, but I suspect an API that allowed work-items to effectively ignore `arrive` and fallback to a regular barrier inside of `wait` would be portable and still permit acceleration when the right hardware is available. @FMarno is trying to do something like this already with work-group copies (see https://github.com/intel/llvm/pull/4950).

Reply:
> but these instructions are specified in terms of how many sub-groups they synchronize rather than how many work-items they synchronize.

I don't think that is an issue for a general use case with all work items in a sub-group either participating or not participating in a barrier. However, it does become an issue when participation in a barrier is not based on just sub-group. I think we can work around this issue by requiring that all subgroups are converged upon calling arrive and wait if independent forward progress is not supported. So I think we could have software implementation based on these two, but I don't see them implemented in libclc.

> I've not thought about this enough, but I suspect an API that allowed work-items to effectively ignore arrive and fallback to a regular barrier inside of wait would be portable and still permit acceleration when the right hardware is available.

Unfortunately that is not exactly the same functionality. Counterexample would be work items in a work group executing arrives for 2 barriers before waiting on these two barriers in different orders (that is possible even without independent forward progress if different orders occur in different sub-groups). This would work with true async barrier and deadlock with your suggestion.

Reply:
> I think we can work around this issue by requiring that all subgroups are converged upon calling arrive and wait if independent forward progress is not supported.

Are you suggesting that all work-items in the sub-group would be required to encounter arrive/wait _and_ that the number of work-items expected would have to be a multiple of the sub-group size?

> Counterexample would be work items in a work group executing arrives for 2 barriers before waiting on these two barriers in different orders (that is possible even without independent forward progress if different orders occur in different sub-groups). This would work with true async barrier and deadlock with your suggestion.

You're right, I hadn't considered corner cases like this.  Have you got an example of what that code would look like?  I'm trying to imagine how likely this pattern is to occur, and how easy it would be to write it accidentally.

If what you said before was broadened to say that all work-items in a work-group must be converged upon calling `wait`, then I think this sort of re-ordering would naturally be disallowed.

Reply:
> Are you suggesting that all work-items in the sub-group would be required to encounter arrive/wait and that the number of work-items expected would have to be a multiple of the sub-group size?

Yes.

> If what you said before was broadened to say that all work-items in a work-group must be converged upon calling wait, then I think this sort of re-ordering would naturally be disallowed.

Yeah, if we add such restriction the trivial implementation is possible. However I am not sure if adding such a limitation makes the async barrier useful at all. With such limitation all use cases can be simplified to use regular group barrier, which would also be more efficient.

Reply:
I think this could get icky very fast for "remainder" subgroups if the subgroup size doesn't evenly divide the work-group size.  If what we need is a subgroup barrier (or even a work-group barrier) it seems easier to describe that directly.

As @pennycook alluded to in comments above and below we have been looking at a very basic asynchronous barrier similar to `cm_sbarrier` that we believe can be supported on all hardware.  It will more or less follow the same convergence rules as the existing barrier but will allow arriving at the barrier separately from waiting at the barrier.  We believe this can provide a performance improvement over an ordinary combined barrier for hardware that has two different phases, but safely degrade to a combined barrier for hardware that does not.

Reply:
OK, I am getting convinced that we can not expose all or most of the features from CUDA in a general extension. So I will change this extension, as suggested in the first comment of this thread, to make it obvious that it is CUDA only.